### ISSUE TITLE
Working around integer limitations

### ISSUE THREAD

**Issue by provinzkraut** [jcrist/msgspec#377]
### Description

Encoding of integers is currently limited to an upper bound of 2^64. I assume this has technical reasons so values fit into a 64 bit integer. However, this may be unexpected for users, since this is not a limitation of the stlib's `json` module. It's also not possible to work around this issue using the `enc_hook`, since it won't be called for natively supported types.

A way to work around this would be highly appreciated!

**Comment by jcrist** [comment]
We could certainly add support for JSON encoding integers that don't fit in an `int64`/`uint64`. This would give the following behavior:

```python
>>> msgspec.json.encode(2**66)
b'73786976294838206464'
```

However, without a bunch of additional work, decoding these will still convert them into floats:

```python
>>> msgspec.json.decode(b'73786976294838206464')
7.378697629483821e+19
```

I'm unlikely to change this in the near future, as it would be a fair amount of work to do without affecting performance.

---

Support for integers larger outside the int64/uint64 range is spotty both in Python and in other languages.

In python, the `json` library supports integers of arbitrary size, but both `ujson`, and `orjson` have the same restrictions as `msgspec` currently. Additionally, `ujson` will error when decoding large integers, while `orjson` and `msgspec` will cast them to floats (`json` parses them as integers).

Outside of Python, Javascript won't accurately roundtrip any integer larger than 53 bits. And JSON implementations in most compiled languages (go, rust, ...) will decode large integers as floats rather than into some BigInt structure by default. Some systems make use of strings for roundtripping large integers to ensure uniform cross-language support, rather than encoding them as ints.

Given the limitations, I'd question the usefulness of supporting them - what systems will be decoding these messages, and how will they be making use of them? See also my comment here: https://github.com/jcrist/msgspec/issues/153#issuecomment-1192813007.

---

Anyway, I'd be happy to add support for encoding large integers in JSON alone. This is a small amount of work, and shouldn't have any negative consequences (performance or otherwise) on the library.

**Comment by provinzkraut** [comment]
> Given the limitations, I'd question the usefulness of supporting them

That's certainly a valid point. The reason this came up with was because we got a bug report over at Litestar (https://github.com/litestar-org/litestar/issues/1544), and because of the combination of msgspec not supporting them natively and also not being able to override the default behaviour, we won't be able to offer users any workarounds, since we strictly handle all JSON encoding with msgspec. 

I'd be perfectly okay with msgspec not supporting this by default if there was a way for users to override this behaviour, even at a performance cost.

If I recall correctly, this would be possible with the new "extension support" (can't seem to remember what you called it), which won't land any time soon?

<hr>

> Anyway, I'd be happy to add support for encoding large integers in JSON alone.

That'd be good enough for me for the time being. It's not a perfect solution but as you outlined, this is an edge case and offers users at least some way to work around this.

**Comment by jcrist** [comment]
> since we strictly handle all JSON encoding with msgspec.

FWIW you would have had the same issue with your previous choice of `orjson` too. Support for this isn't common, but at least on the encoding side should be pretty straightforward to add.

Anyway, I should be able to fix this in the next couple days. Happy to help.

**Comment by jcrist** [comment]
Was able to hack something together during a meeting - should be fixed by #378.

**Comment by provinzkraut** [comment]
Amazing! @jcrist
